spec.bs

```diff
@@ -1387,6 +1478,227 @@ To <dfn>build an interest group passed to generateBid</dfn> given an [=interest
+  1. If |topLevelAuctionConfig| is null:
+    1. If |parsedAdditionalBid|["topLevelSeller"] [=map/exists=], then return failure.
+  1. Otherwise:
+    1. If |parsedAdditionalBid|["topLevelSeller"] does not [=map/exist=], then return failure.
+    1. Let |bidTopLevelSeller| be the result of running the [=parse an https origin=] with
+      |parsedAdditionalBid|["topLevelSeller"].
+    1. If |bidTopLevelSeller| is failure, or |bidTopLevelSeller| is not [=same origin=] with
+      |topLevelAuctionConfig|'s [=auction config/seller=], then return failure.
+  1. If |parsedAdditionalBid|["interestGroup"] does not [=map/exist=], then return failure.
+  1. Let |igMap| be |parsedAdditionalBid|["interestGroup"].
+  1. Return failure if any the following conditions hold:
+    * |igMap| is not a [=map=];
+    * |igMap|["name"] does not [=map/exist=], or it's not a [=string=];
+    * |igMap|["biddingLogicURL"] does not [=map/exist=], or it's not a [=string=];
+    * |igMap|["owner"] does not [=map/exist=], or it's not a [=string=];
+  1. Let |igOwner| be the result of running [=parse an https origin=] given |igMap|["owner"].
+  1. Let |igName| be |igMap|["name"].
+  1. Let |igBiddingUrl| be the result of running [=url parser=] on |igMap|["biddingLogicURL"].
+  1. Return failure if any of the following conditions hold:
+    * |igOwner| is failure;
+    * |auctionConfig|'s [=auction config/interest group buyers=] does not [=list/contain=] |igOwner|;
+    * |igBiddingUrl| is failure;
+    * |igOwner| is not [=same origin=] with |igBiddingUrl|.
+  1. Let |ig| be a new [=interest group=] with the following properties:
+    :   [=interest group/owner=]
+    ::  |igOwner|
+    :   [=interest group/name=]
+    ::  |igName|
+    :   [=interest group/bidding url=]
+    ::  |igBiddingUrl|
+  1. If |parsedAdditionalBid|["bid"] does not [=map/exist=], or it's not a [=map=], return failure.
+  1. Let |bidMap| be |parsedAdditionalBid|["bid"].
+  1. If |bidMap|["render"] does not [=map/exist=] or it's not a [=string=], then return failure.
+  1. Let |renderUrl| be the result of running [=url parser=] on |bidMap|["render"].
+  1. If |renderUrl| is failure, then return failure.
+  1. Let |ad| be a new [=interest group ad=] whose [=interest group ad/render url=] is |renderUrl|.
+  1. Set |ig|'s [=interest group/ads=] to « |ad| ».
+  1. Let |bidVal| be |bidMap|["bid"] if it [=map/exists=], otherwise return failure.
+  1. If |bidVal| is not a {{double}}, or it's less than or equal to 0, then return failure.
+  1. Let |adMetadata| be "null".
+  1. If |bidMap|["ad"] [=map/exists=]:
+    1. Set |adMetadata| to the result of running [=serialize an Infra value to a JSON string=] with
+      |bidMap|["ad"].
+  1. Let |bidCurrency| be null.
+  1. If |bidMap|["bidCurrency"] [=map/exists=]:
+    1. If |bidMap|["bidCurrency"] is not a [=string=], or the result of [=checking whether a string
+      is a valid currency tag=] is failure:
+      1. Return failure.

```

Comment:
I'll stop writing this comment for now, but I'd recommend going through and taking some of these really small if bodies and just moving them up a line, so they're not completely nested. You can use your own judgement, but in general the spec seems a little over-wrapped IMO!